### DISCUSSION TITLE
What does max_upload_speed mean?

### DISCUSSION THREAD

**Discussion by staticgc** [sagebind/isahc#298] [Questions] [answered]
There is a configuration [max_upload_speed](https://docs.rs/isahc/1.1.0/isahc/config/trait.Configurable.html#method.max_upload_speed) which seems to suggest that it controls bytes/sec kind of rate limiting.

How to interpret this for HTTP/1 & HTTP/2?

If HTTP/1 then does it mean the limit apply for all connections per [Client](https://docs.rs/isahc/1.1.0/isahc/struct.HttpClient.html) instance?

If HTTP/2 then does it mean the rate limit is per Body of the request? So if limit is 10 mbps for each request the limit is 10 mbps?

**Comment by sagebind** [comment, marked as answer]
The docs say:

> Set a maximum upload speed for the request body, in bytes per second.

This isn't as clear as it could be, but what this means is that this limit is applied to each request body upload individually. For example, if you had three concurrent requests and your client was configured with a 10 MiB/s max upload speed, then effectively your program might cumulatively use up to 30 MiB/s, 10 for each request. Setting a max upload speed on a `Client` instance is just a way of setting a default value for each request it sends.

The use of HTTP/1.x or HTTP/2 has no effect on the behavior, nor does connection pooling or reuse. The limit is always applied to individual requests.

I think it might be a good idea to improve the docs here to be more clear.
